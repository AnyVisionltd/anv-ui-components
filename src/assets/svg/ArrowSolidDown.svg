<svg xmlns="http://www.w3.org/2000/svg" id="prefix__Arrow_solid_down" width="24" height="24" data-name="Arrow solid down" viewBox="0 0 24 24">
    <path id="prefix__Shape" d="M0 0H24V24H0z" style="fill:#02234b;opacity:0"/>
    <path id="prefix__Arrow_solid_down-2" d="M5.247.86a1 1 0 0 1 1.505 0l3.8 4.339A1 1 0 0 1 9.8 6.857H2.2A1 1 0 0 1 1.451 5.2z" data-name="Arrow solid down" transform="rotate(180 9 8.116)"/>
</svg>
