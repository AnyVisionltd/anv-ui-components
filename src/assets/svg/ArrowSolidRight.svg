<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">
    <path d="M5.247.86a1 1 0 0 1 1.505 0l3.8 4.339A1 1 0 0 1 9.8 6.857H2.2A1 1 0 0 1 1.451 5.2z" transform="rotate(90 5.088 11.09)"/>
</svg>
