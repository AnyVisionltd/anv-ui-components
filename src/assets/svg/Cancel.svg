<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16">
    <g id="prefix__Cancel" transform="translate(-82 -8)">
        <path id="prefix__Shape" d="M0 0H16V16H0z" transform="translate(82 8)" style="opacity:0"/>
        <path id="prefix__Path_2223" d="M12.13 11l2.634-2.634a.8.8 0 1 0-1.131-1.131L11 9.868 8.365 7.234a.8.8 0 1 0-1.131 1.131L9.868 11l-2.634 2.633a.8.8 0 1 0 1.131 1.131L11 12.13l2.634 2.634a.8.8 0 1 0 1.131-1.131z" data-name="Path 2223" transform="translate(79.001 5.001)"/>
    </g>
</svg>
