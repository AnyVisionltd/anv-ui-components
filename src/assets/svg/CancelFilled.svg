<svg xmlns="http://www.w3.org/2000/svg" id="prefix__Cancel_X" width="24" height="24" data-name="Cancel X" viewBox="0 0 24 24">
    <path id="prefix__Shape" d="M0 0H24V24H0z" style="opacity:0;fill:#02234b"/>
    <path id="prefix__Subtraction_5" d="M8 16A8 8 0 0 1 2.344 2.343a8 8 0 1 1 11.312 11.31A7.941 7.941 0 0 1 8 16zm0-6.871l2.633 2.634a.841.841 0 0 0 .316.2.806.806 0 0 0 .251.037.791.791 0 0 0 .564-.234A.8.8 0 0 0 12 11.2a.793.793 0 0 0-.234-.564L9.131 8l2.633-2.634a.8.8 0 0 0 0-1.131.8.8 0 0 0-1.131 0L8 6.87 5.365 4.236a.8.8 0 0 0-1.131 0 .8.8 0 0 0 0 1.131L6.869 8l-2.635 2.635A.792.792 0 0 0 4 11.2a.8.8 0 0 0 .8.8.791.791 0 0 0 .564-.234L8 9.131z" data-name="Subtraction 5" transform="translate(4 4)"/>
</svg>
