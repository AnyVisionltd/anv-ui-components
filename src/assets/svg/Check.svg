<svg xmlns="http://www.w3.org/2000/svg" id="prefix__check" width="24" height="24" viewBox="0 0 24 24">
    <path id="prefix__Shape" d="M0 0H24V24H0z" style="opacity:0;fill:#00d0a0"/>
    <path id="prefix__check-2" d="M18.169 7.2a.961.961 0 0 0-1.364 0l-7.157 7.164-3.006-3.016a.982.982 0 1 0-1.364 1.412l3.689 3.689a.961.961 0 0 0 1.364 0l7.838-7.839a.961.961 0 0 0 0-1.412z" data-name="check" transform="translate(.272 -.163)" />
</svg>
