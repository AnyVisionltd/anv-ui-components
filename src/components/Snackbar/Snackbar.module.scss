@import "~@anyvision/style-guide";

.snackbar {
  display: none;
  min-width: av-size(sz-480);
  height: 64px;
  color: av-color(primary);
  background-color: av-color(surface);
  align-items: center;
  border-left: solid 4px currentColor;
  border-radius: 0 av-radius() av-radius() 0 ;
  padding: 0 16px;
  justify-content: space-between;
}

.isOpen {
  display: flex;
}

.messageContainer {
  display: flex;
  align-items: center;

  .message {
    color: av-color(content);
  }

  .leadingIcon {
    color: currentColor;
    fill: currentColor;
    margin-right: av-space(tiny);
  }
}

.actionsContainer {
  display: flex;

  .actionButton {
    color: currentColor;
  }

  .trailingIcon {
    margin-left: av-space(tiny);
  }
}
