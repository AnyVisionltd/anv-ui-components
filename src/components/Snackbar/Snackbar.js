import React, { useEffect, useCallback } from 'react'
import classNames from 'classnames'
import propTypes from 'prop-types'
import { Button, IconButton, Portal } from '../../index'
import { ReactComponent as CloseIcon } from '../../assets/svg/Close.svg'
import styles from './Snackbar.module.scss'

const Snackbar = ({
  message,
  actionText,
  leadingIcon,
  closeIcon,
  open,
  className,
  onOpen,
  onClose,
  hideTimeout,
  onAction,
}) => {
  const timerHide = React.useRef()

  const setHideTimeout = useCallback(() => {
    if (hideTimeout === null) {
      return
    }

    clearTimeout(timerHide.current)
    timerHide.current = setTimeout(() => {
      onClose()
    }, hideTimeout)
  }, [hideTimeout, onClose])

  useEffect(() => {
    if (open) {
      onOpen()
      setHideTimeout()
    }

    return () => {
      clearTimeout(timerHide.current)
    }
  }, [open, hideTimeout, setHideTimeout, onOpen])

  const classes = classNames(
    styles.snackbar,
    className,
  )

  return !open ? null : (
    <Portal
      containerId="snackbar-portal"
      className={ styles.portal }
    >
      <div
        className={ classes }
        onMouseEnter={ () => clearTimeout(timerHide.current) }
        onMouseLeave={ setHideTimeout }
      >
        <div className={ styles.messageContainer }>
          {
            leadingIcon
              ? (
                <span className={ styles.leadingIcon }>
                  { leadingIcon }
                </span>
              )
              : null
          }
          <span className={ styles.message }>
            { message }
          </span>
        </div>
        <div className={ styles.actionsContainer }>
          { actionText
            ? (
              <Button
                variant="ghost"
                size="small"
                onClick={ onAction }
              >
                { actionText }
              </Button>
            )
            : null }
          {
            closeIcon
              ? (
                <IconButton
                  variant="ghost"
                  className={ styles.closeIcon }
                  onClick={ onClose }
                >
                  { closeIcon }
                </IconButton>
              )
              : null
          }
        </div>
      </div>
    </Portal>
  )
}

Snackbar.defaultProps = {
  closeIcon: <CloseIcon />,
  onOpen: () => {},
  onClose: () => {},
  hideTimeout: 5000,
  onAction: () => {},
}

Snackbar.propTypes = {
  /** The message to display. */
  message: propTypes.string.isRequired,
  /** The action text. */
  actionText: propTypes.string,
  /** The icon before the message.  */
  leadingIcon: propTypes.element,
  /** The icon trailing icon, used for close. Set to false for remove */
  closeIcon: propTypes.oneOfType([propTypes.element, propTypes.bool]),
  /** If <code>true</code> display the snackbar. */
  open: propTypes.bool,
  /**
   *  The number of milliseconds to wait before automatically calling
   *  the <code>onClose</code> function.
   *  <code>onClose</code> should then set the state of the open prop to hide the Snackbar.
   *  Disable this behavior by <code>null</code> value.
   *  */
  hideTimeout: propTypes.number,
  /** Callback fired when the component opened. */
  onOpen: propTypes.func,
  /**
   * Callback fired when the component requests to be closed.
   * Typically onClose is used to set state in the parent component,
   * which is used to control the Snackbar open prop.
   * */
  onClose: propTypes.func,
  /** Callback fired when <code>closeIcon</code> click or after hideTimeout */
  onAction: propTypes.func,
  /** For css customization. */
  className: propTypes.string,
}

export default Snackbar
