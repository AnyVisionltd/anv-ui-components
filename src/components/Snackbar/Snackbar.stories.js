import React, { useState, useCallback } from 'react'
import Snackbar from './Snackbar'
import { ReactComponent as SunIcon } from '../../assets/svg/Sun.svg'
import { ReactComponent as CheckIcon } from '../../assets/svg/Check.svg'
import { centerDecorator } from '../../utils/storybook/decorators'
import styles from '../../styles/storybook/index.module.scss'
import { Button } from '../Button'

export default {
	title: 'Components/Snackbar',
	component: Snackbar,
	decorators: [centerDecorator],
}

export const Default = () => {
	const [isOpen, setIsOpen] = useState(false)
	return (
		<>
			<Button onClick={ () => setIsOpen(true) }>Open</Button>
			<Snackbar
				isOpen={ isOpen }
				onClose={ useCallback(() => setIsOpen(false), []) }
				message="This Is The Message"
			/>
		</>
	)
}

export const LeadingIcon = () => {
	const [isOpen, setIsOpen] = useState(false)
	return (
		<>
			<Button onClick={ () => setIsOpen(true) }>With Leading Icon</Button>
			<Snackbar
				isOpen={ isOpen }
				onClose={ useCallback(() => setIsOpen(false), []) }
				message="This Is The Message"
				leadingIcon={ <SunIcon /> }
			/>
		</>
	)
}

export const CloseIcon = () => {
	const [isOpenDefault, setIsOpenDefault] = useState(false)
	const [isOpenCustom, setIsOpenCustom] = useState(false)
	const [isOpenWithout, setIsOpenWithout] = useState(false)
	return (
		<>
			<div className={ styles.marginFlexContainer }>
				<Button onClick={ () => setIsOpenDefault(true) }>Default Icon</Button>
				<Button onClick={ () => setIsOpenCustom(true) }>Custom Icon</Button>
				<Button onClick={ () => setIsOpenWithout(true) }>Without Icon</Button>
			</div>
			<Snackbar
				isOpen={ isOpenDefault }
				onClose={ useCallback(() => setIsOpenDefault(false), []) }
				message="Default trailing icon"
			/>
			<Snackbar
				isOpen={ isOpenCustom }
				onClose={ useCallback(() => setIsOpenCustom(false), []) }
				message="Custom trailing icon"
				closeIcon={ <CheckIcon /> }
			/>
			<Snackbar
				isOpen={ isOpenWithout }
				onClose={ useCallback(() => setIsOpenWithout(false), []) }
				message="Without trailing icon"
				closeIcon={ false }
			/>
		</>
	)
}

export const Action = () => {
	const [isOpen, setIsOpen] = useState(false)
	return (
		<>
			<Button onClick={ () => setIsOpen(true) }>With Action</Button>
			<Snackbar
				isOpen={ isOpen }
				onClose={ useCallback(() => setIsOpen(false), []) }
				message="This Is The Message"
				action={ <Button variant="ghost" size="small"> undo </Button> }
			/>
		</>
	)
}
