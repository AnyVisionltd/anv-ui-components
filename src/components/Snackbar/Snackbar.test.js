import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import Snackbar from './Snackbar'
import { ReactComponent as SunIcon } from '../../assets/svg/Sun.svg'

describe('<Snackbar />', () => {
  describe('Render Snapshots', () => {
    it('should render snackbar when isOpen true', () => {
      const { container } = render(<Snackbar message='test message' isOpen />)
      expect(container).toMatchSnapshot()
    })

    it('should NOT render snackbar when isOpen false', () => {
      const { container } = render(
        <Snackbar message='test message' isOpen={false} />,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render leadingIcon', () => {
      const { container } = render(
        <Snackbar message='test message' isOpen leadingIcon={<SunIcon />} />,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render custom closeIcon', () => {
      const { container } = render(
        <Snackbar message='test message' isOpen closeIcon={<SunIcon />} />,
      )
      expect(container).toMatchSnapshot()
    })

    it('should NOT render closeIcon', () => {
      const { container } = render(
        <Snackbar message='test message' isOpen closeIcon={false} />,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render action', () => {
      const { container } = render(
        <Snackbar message='test message' isOpen action='action' />,
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('Closing events', () => {
    it('should fire onClose when click closeButton', () => {
      const onClose = jest.fn()
      const { getByRole } = render(
        <Snackbar message='test message' isOpen onClose={onClose} />,
      )
      const node = getByRole('button')
      fireEvent.click(node)
      expect(onClose).toHaveBeenCalledTimes(1)
    })

    it('should fire onClose after hideTimeout', async () => {
      const onClose = jest.fn()
      render(
        <Snackbar
          message='test message'
          isOpen
          onClose={onClose}
          hideTimeout={1}
        />,
      )
      await waitFor(() => expect(onClose).toHaveBeenCalledTimes(1))
    })

    it('should NOT fire onClose after hideTimeout', async () => {
      const onClose = jest.fn()
      const { container } = render(
        <Snackbar
          message='test message'
          isOpen
          onClose={onClose}
          hideTimeout={1}
        />,
      )
      const node = container.firstChild.firstChild
      fireEvent.mouseEnter(node)
      await waitFor(() => expect(onClose).toHaveBeenCalledTimes(0))
      fireEvent.mouseLeave(node)
      await waitFor(() => expect(onClose).toHaveBeenCalledTimes(1))
    })
  })
})
