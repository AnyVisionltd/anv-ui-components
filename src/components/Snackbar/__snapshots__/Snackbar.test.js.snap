// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Snackbar /> Render Snapshots should NOT render closeIcon 1`] = `
<div>
  <portal-for-test>
    <div
      class="snackbar"
    >
      <div
        class="messageContainer"
      >
        <span>
          test message
        </span>
      </div>
      <div
        class="actionsContainer"
      />
    </div>
  </portal-for-test>
</div>
`;

exports[`<Snackbar /> Render Snapshots should NOT render snackbar when isOpen false 1`] = `<div />`;

exports[`<Snackbar /> Render Snapshots should render action 1`] = `
<div>
  <portal-for-test>
    <div
      class="snackbar"
    >
      <div
        class="messageContainer"
      >
        <span>
          test message
        </span>
      </div>
      <div
        class="actionsContainer"
      >
        action
        <button
          class="button large ghost iconButton small ghost closeIcon"
          type="button"
        >
          <svg />
        </button>
      </div>
    </div>
  </portal-for-test>
</div>
`;

exports[`<Snackbar /> Render Snapshots should render custom closeIcon 1`] = `
<div>
  <portal-for-test>
    <div
      class="snackbar"
    >
      <div
        class="messageContainer"
      >
        <span>
          test message
        </span>
      </div>
      <div
        class="actionsContainer"
      >
        <button
          class="button large ghost iconButton small ghost closeIcon"
          type="button"
        >
          <svg />
        </button>
      </div>
    </div>
  </portal-for-test>
</div>
`;

exports[`<Snackbar /> Render Snapshots should render leadingIcon 1`] = `
<div>
  <portal-for-test>
    <div
      class="snackbar"
    >
      <div
        class="messageContainer"
      >
        <span
          class="leadingIcon"
        >
          <svg />
        </span>
        <span>
          test message
        </span>
      </div>
      <div
        class="actionsContainer"
      >
        <button
          class="button large ghost iconButton small ghost closeIcon"
          type="button"
        >
          <svg />
        </button>
      </div>
    </div>
  </portal-for-test>
</div>
`;

exports[`<Snackbar /> Render Snapshots should render snackbar when isOpen true 1`] = `
<div>
  <portal-for-test>
    <div
      class="snackbar"
    >
      <div
        class="messageContainer"
      >
        <span>
          test message
        </span>
      </div>
      <div
        class="actionsContainer"
      >
        <button
          class="button large ghost iconButton small ghost closeIcon"
          type="button"
        >
          <svg />
        </button>
      </div>
    </div>
  </portal-for-test>
</div>
`;
